proof/invariant-abstract/IpcDet_AI.thy

```diff
@@ -1550,34 +1543,20 @@ lemma reply_push_invs_helper:
     apply (wpsimp wp: set_sc_replies_valid_replies update_sched_context_valid_idle)
    apply (wpsimp simp: get_simple_ko_def get_object_def
                    wp: valid_sc_typ_list_all_reply valid_ioports_lift)
-  apply (clarsimp split: if_splits, intro conjI impI)
-                apply (clarsimp simp: is_reply obj_at_def is_sc_obj_def)+
-        apply (subgoal_tac "reply_sc_reply_at (\<lambda>sp. sp = Some sc_ptr) reply_ptr s")
-         apply (clarsimp simp:reply_sc_reply_at_def obj_at_def)
-        apply (fastforce simp: sym_refs_reply_sc_reply_at invs_sym_refs reply_at_ppred_reply_at)
-       apply (clarsimp simp: is_reply obj_at_def is_sc_obj_def)+
-  apply (clarsimp simp: invs_def valid_state_def valid_pspace_def sc_replies_sc_at_def
-                        valid_obj_def valid_sched_context_def
-                 elim!: obj_at_valid_objsE)
-  apply (drule_tac s="a # list" in sym, simp)
-  apply (rule conjI, clarsimp simp: obj_at_def is_reply)
-  apply (rule conjI, clarsimp simp: replies_with_sc_def image_def sc_replies_sc_at_def obj_at_def)
-   apply (subgoal_tac "reply_ptr \<in>  set (sc_replies sc)")
-    apply (drule_tac x=sc_ptr in spec, fastforce)
-   apply fastforce
-  apply (rule conjI, rule replies_with_sc_upd_replies_valid_replies_add_one, clarsimp)
+  apply (subgoal_tac "list_all (\<lambda>r. reply_at r s) (a # list) \<and> reply_ptr \<notin> set (a # list) \<and> distinct (a # list)")
+  apply (clarsimp simp: invs_def valid_pspace_def valid_state_def)
```

Comment:
Doesn't the `subgoal_tac` generate another subgoal?

Reply:
Yes, there should be and additional indent here